compiler/daml-lf-optimize/src/DA/Daml/LF/Optimize.hs

```diff
@@ -0,0 +1,621 @@
+        ... v E2 ...                    --> (\(v:T) . E1) E2
+
+The optimizer works in the manner of an evaluator, with `SemValue` the type of the resulting
+_semantic value _ of the _evaluation_. The implementation proceeds via the interplay of two
+functions: `reflect` and `reify`.
+
+        reflect :: LF.Expr -> Effect SemValue
+        reify :: SemValue -> Effect LF.Expr
+
+Both transformations take place in the `Effect` monad which manages context information, and state
+required for fresh name generation. (And the current packageId, but that might not be necessary).
+
+The type of (LF)expressions is embedded within the type of `SemValue`, via the `Syntax`
+constructor. Each use of `Syntax` represents a case where we are giving up attempting to normalise
+further.
+
+        data SemValue = Syntax LF.Expr | ...
+
+The other constructors: `Macro`, `Struct` and `TyMacro` correspond to the three kinds of redex we
+which to handle specially. And for each there is a corresponding _apply_ function:
+
+        termApply :: (SemValue,SemValue) -> Effect SemValue
+        typeApply :: SemValue -> LF.Type -> Effect SemValue
+        applyProjection :: LF.FieldName -> SemValue -> Effect SemValue
+
+This is where the redex reduction is actually performed when possible.  When the reduction is not
+possible (because the applicant is `Syntax` rather than the desired special form) then the
+Syntactic-level application is reconstructed, reifying any normalized expression back to syntactic
+LF expressions as required.
+
+This code is just proof-of-concept/WIP, and not all forms of LF are yet handled. In particular it
+currently only performs optimization of expressions bound as top level definitions.
+
+The approach to (mutual)recursion is to track which top level definitions we have already expanded,
+and not recursively expand a definition within the scope of its own expansion. See `WithDontInline`
+
+
+Here is a (very) small example of the effect of optimizing a DAML function to decrement an integer:
+with the appoximate DAML-LF shown (with nicer vars names for explication!)
+
+    DAML:
+        decrement x = x - 1
+
+    DAML-LF (original):
+
+        def decrement = \(x:Int64) -> (-)[Int64] $fAdditiveInt x 1
+
+        def (-) = \[t].(\(dict:GHC.Num:Additive t) -> dict.m_- ())
+
+        def $fAdditiveInt =
+          { m_+      = \(_:Unit) -> ADDI
+          , m_aunit  = \(_:Unit) -> $caunit
+          , m_-      = \(_:Unit) -> SUBI
+          , m_negate = \(_:Unit) -> $cnegate
+          }
+
+        def $caunit = 0
+
+        def $cnegate = \(x:Int64) -> (-)[Int64] $fAdditiveInt 0 x
+
+
+    DAML-LF (optimized):
+
+        def decrement = \(_v:Int64) -> SUBI _v 1
+
+
+One potential problem with this approach to optimization is blow up in the size of the generated
+code. So far I haven't witnessed this, but I have only run quite small examples.
+
+Potential speedups.. Using the classic "nfib" function DAML:
+
+    nfib : Int -> Int
```

Comment:
A comparison to what seems like an optimal handcrafted DALF would also be interesting here.

Reply:
How would you beat the output from the optimizer?

    @location(19:0-19:4)
    def nfib : Int64 -> Int64 =
      \(_v268 : Int64).
        case EQUAL @Int64 _v268 0 of
          false ->
            case EQUAL @Int64 _v268 1 of
              false ->
                ADD_INT64
                  (ADD_INT64
                     (Main:nfib (SUB_INT64 _v268 1))
                     (Main:nfib (SUB_INT64 _v268 2)))
                  1
              true -> 1
          true -> 1

Reply:
That does indeed look pretty optimal :+1: